src/main/java/io/vertx/core/net/impl/pool/NonBlockingSynchronization1.java

```diff
@@ -0,0 +1,36 @@
+package io.vertx.core.net.impl.pool;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class NonBlockingSynchronization1<S> implements Synchronization<S> {
+
+  private final ConcurrentLinkedDeque<Action<S>> q = new ConcurrentLinkedDeque<>();
+  private final AtomicInteger s = new AtomicInteger();
+  private final S state;
+
+  public NonBlockingSynchronization1(S state) {
+    this.state = state;
+  }
+
+  @Override
+  public void execute(Action<S> action) {
+    q.add(action);
+    if (s.incrementAndGet() == 1) {
```

Comment:
I would do instead:
```java
      if (s.get() > 0) {
         return;
      }
      while (s.compareAndSet(0, 1)) {
         try {
            Action<S> a;
            while ((a = q.poll()) != null) {
               Runnable post = a.execute(state);
               if (post != null) {
                  post.run();
               }
            }
         } finally {
            s.set(0);
            // full barrier here:
            // the q emptiness check cannot happen before 
            if (q.isEmpty()) {
               return;
            }
         }
      }
```
This should never let any actions to be dropped, awaiting to be executed forever, *but* it has another issue: fairness.

If other threads keep on submitting actions the risk is that a single thread will take the whole execution hit without completing, leading to a huge time occupying its own `Thread`.
If the executor is running on the event loop it means that no other tasks will be executed on that thread..

Reply:
thanks.

the fairness si fine because it is for a pool with a bound number of requests to be performed. worth case I would set a threshold and when too many actions are performed then a task is executed on the event loop to continue processing.

Reply:
Note that I think we will not keep this implementation but instead use the `NonBlockingSynchronization2` that we discussed, perhaps you can have a look as well ?